### PR TITLE
Change the DPAS layout mapping in advance for migrating the interface from GenISA to OpenCL

### PR DESCRIPTION
**Pull request by chengjunlu**
The OpenCL interface only support to pack A operand to opaque i16 for scalar type bits width <=16. Change the DPAS layout mapping in advance for migrating the interface from GenISA to OpenCL.

### CODE REVIEW

**Review by whitneywhtsang — APPROVED** [review]

File: lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 1]
```diff
@@ -218,16 +218,20 @@ DpasEncodingAttr::getShapePerCTATileForDotOperands(ArrayRef<int64_t> shape,
...
-    auto shapeA = getShapeA();
-    auto subGroupSize = getSubGroupSize();
-    auto systolicDepth = getSystolicDepth();
-    if (subGroupSize < systolicDepth) {
+    SmallVector<unsigned> shapeA = getShapeA();
+    unsigned subGroupSize = getSubGroupSize();
+    unsigned opsPerChannel = getOpsPerChannel();
+
+    // pack the value to i16 for scalar bit width <=16.
+    unsigned packedOpsPerLane = opsPerChannel == 4 ? 2 : 1;
```
Comment (etiotto) [inline lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:226]: opsPerChannel == 4 is used for elem. type FP8.  What about the case when operands are less than 8 bit precision? In that case opsPerChannel is 8.  Do we need to pack them ?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 2]
```diff
@@ -183,12 +183,19 @@ static LLVM::CallOp createGenISADPAS(TritonGEN::MatrixDPASOp op,
   IntegerType int32Ty = rewriter.getIntegerType(32);
 
   Value a = op.getA();
-  auto aTy = VectorType::get(op.getRc(), int16Ty);
+  auto aOrigTy = cast<VectorType>(a.getType());
+  auto bitWidth = aOrigTy.getNumElements() *
+                  aOrigTy.getElementType().getIntOrFloatBitWidth();
+  auto aTy = VectorType::get(bitWidth / 16, int16Ty);
   if (a.getType() != aTy)
```
Comment (etiotto) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:190]:
```suggestion
  if (aOrigTy != aTy)
```

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 3]
```diff
@@ -183,12 +183,19 @@ static LLVM::CallOp createGenISADPAS(TritonGEN::MatrixDPASOp op,
   IntegerType int32Ty = rewriter.getIntegerType(32);
 
   Value a = op.getA();
-  auto aTy = VectorType::get(op.getRc(), int16Ty);
+  auto aOrigTy = cast<VectorType>(a.getType());
+  auto bitWidth = aOrigTy.getNumElements() *
```
Comment (etiotto) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:187]: auto -> unsigned

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 4]
```diff
@@ -183,12 +183,19 @@ static LLVM::CallOp createGenISADPAS(TritonGEN::MatrixDPASOp op,
...
     a = rewriter.create<LLVM::BitcastOp>(loc, aTy, a);
 
   Value b = op.getB();
-  auto bTy = VectorType::get(8, int32Ty);
+
+  auto bOrigTy = cast<VectorType>(b.getType());
+  bitWidth = bOrigTy.getNumElements() *
+             bOrigTy.getElementType().getIntOrFloatBitWidth();
+  auto bTy = VectorType::get(bitWidth / 32, int32Ty);
   if (b.getType() != bTy)
```
Comment (etiotto) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:199]:
```suggestion
  if (bOrigTy != bTy)
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp [hunk 5]
```diff
@@ -86,14 +86,19 @@ DpasMatmulLoader<opIdx>::computeLdsMatOffs(Value warpId, Value laneId,
...
 
   Value laneRowIndex, laneColIndex;
-  unsigned repRowsPerInst, rowsPerWarp;
+  unsigned repRowsPerInst, rowsPerWarp, repOpsPerRow;
   switch (opIdx) {
   case 0: {
-    rowsPerWarp = threadsPerWarp / systolicDepth;
+    SmallVector<unsigned> shapeA = dpasLayout.getShapeA();
+    // Unlike the operand B, to pack the value to i16 for scalar bit width <=16.
+    unsigned packedOpsPerLane = opsPerChannel == 4 ? 2 : 1;
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp:94]: When opsPerChannel is 8 should we pack the operands ?

File: lib/Dialect/TritonIntelGPU/IR/Dialect.cpp (hunk 1 above)
Comment (chengjunlu) [reply lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:226]: Haven't considered the operands type of u4, s4, u2, s2 yet.
As the OpenCL interface the A operands is i16 opaque, maybe `ceil(opsPerChannel, 2)` is more general.
How about to change this along with the u4, s4, u2, s2 data type supporting?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 2 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp]: Changed

File: lib/Dialect/TritonIntelGPU/IR/Dialect.cpp (hunk 1 above)
Comment (chengjunlu) [reply lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:228]: I added a assertion to protect the code from invalid opsPerChannel. The Dpas verifier also protect from invalid opsPerChannel.

```
  if (!(opsPerChan == 1 || opsPerChan == 2 || opsPerChan == 4)) {
    return emitError() << "opsPerChannel must be 1, 2 or 4, but was:"
                       << opsPerChan;
  }
```

Let's add the opsPerChannel=8 support along with the u4, s4, u2, s2 supporting.

Comment (etiotto) [reply lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:228]: ok let's do it in a separate PR. Is good you have the assertion for now.

File: third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp (hunk 5 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandDPAS.cpp:96]: see comments above, for now we will assert when opsPerChannel is 8

**Review by etiotto — APPROVED** [review]

**Pull request merged**